core/blockchain/blockchain_gateway.py

```diff
@@ -0,0 +1,355 @@
+        Return weights after training terminates
+        TODO: add condition to check if training for specific model terminated
+        """
+        final_weights = self.getter(header)
+        return final_weights
+
+    ##########################################################################
+    ###                          PROVIDER SECTION                          ###
+    ##########################################################################
+
+    async def start_listening(self, event_filter, handler, poll_interval=5):
+        """
+        Starts an indefinite loop that listens for a specific event to occur.
+        Called in `filter_set`. Filters are some condition that must be
+        fulfilled on a per tx basis
+        `poll_interval` specifies the number of seconds to stall before each
+        poll
+        """
+        while True:
+            filtered_diffs = self.get_state_diffs(event_filter)
+            if filtered_diffs:
+                return filtered_diffs
+            await asyncio.sleep(poll_interval)
+
+    def filter_set(self, event_filter, handler):
+        """
+        Calls async method `start_listening` and called by various listening
+        methods
+        """
+        asyncio.set_event_loop(asyncio.new_event_loop())
+        loop = asyncio.get_event_loop()
+        try:
+            filtered_diffs = loop.run_until_complete(
+                self.start_listening(event_filter, handler))
+            check = [handler(diff) for diff in filtered_diffs]
+        finally:
+            loop.close()
+        return check
+
+    def handle_decentralized_learning_trainer(self, tx: dict) -> None:
+        """
+        This callback will be triggered by the Listener if it finds the 
+        method signature it's looking for.
+        The parameters (model weights, model config) will be downloaded 
+        and put into the optimizer initially. So the optimizer knows this info.
+        """
+        logging.info("handling decentralized learning... {}".format(tx))
+        key = tx.get('key')
+        value = tx.get('content')
+        args = {'key': key, 'content': self._ipfs_to_content(value)}
+        self.communication_manager.inform("new_session", args)
+        self.listen_new_weights()
+
+    def handle_new_weights(self, tx: dict):
+        """
+        handle_new_weights() method downloads weights and does something with
+        it. This callback is triggered by the Listener when it sees new weights 
+        intended for its node ID. Downloads the weights, and gives them to the
+        comm. mgr which gives them to the relevant optimizer 
+        -which should do the moving average.
+        """
+        logging.info("handling new weights...{}".format(tx))
+        key = tx.get('key')
+        value = tx.get('content')
+        args = {'key': key, 'content': self._ipfs_to_content(value)}
+        # TODO: Put into in-memory datastore.   
+        self.communication_manager.inform("new_weights", args)
```

Comment:
"new_weights" should be uppercase for consistency. However, ideally you should use the enum here, as well as in all other places instead of the magic strings.

Reply:
wasn't sure whether we were using enum or strings actually, will update

Reply:
@nzoghb to clarify, didn't do above, but you should